sources/models/src/modeled_types/shared.rs

```diff
@@ -349,3 +349,77 @@ mod test_version {
         }
     }
 }
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// DNSDomain represents a string that is a valid DNS domain. It stores the
+/// original string and makes it accessible through standard traits. Its purpose
+/// is input validation, for example validating the kubelet's "clusterDomain"
+/// config value.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct DNSDomain {
+    inner: String,
+}
+
+impl TryFrom<&str> for DNSDomain {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        ensure!(input.len() <= 253, error::InvalidDomainName { input });
+
+        let labels = input.split(".");
+        for (i, label) in labels.enumerate() {
+            ensure!(i < 127, error::InvalidDomainName { input });
+            ensure!(label.len() <= 63, error::InvalidDomainName { input });
+            let last = label.len() - 1;
+            for (j, c) in label.chars().enumerate() {
+                ensure!(
+                    (j != 0 && j != last && (c.is_alphanumeric() || c == '-'))
```

Comment:
I'm having trouble understanding this bit, how does this work?  It seems like `j` will be 0 in the first iteration, since `enumerate` starts at 0, and that would short-circuit the `j != 0` line to be false, and the `ensure` would fail... same when we hit `j == last`.  What am I missing?  :)

I was naively thinking we'd need separate checks, like:
```rust
if j == 0 || j == last {
    ensure!(c.is_alphanumeric(), error::bla);
} else {
    ensure!(c.is_alphanumeric() || c == '-', error::bla);
}
```

Maybe something like that would be more obvious to readers in any case?

Reply:
Yup. Not very clear.  That code will be removed with the switch to url::Host::parse.

For what it's worth... There were two lines in that || condition. If that first line that you mentioned was false, then the second would have to be true.